File: 5. Semester/Deklarativ Programmering/TØ/Week 8/nat.pl
nat(z).
nat(suc(X)) :- nat(X).

add(X, z, X).
add(X, suc(W), suc(N)) :- add(X, W, N).

numc(0, z).
numc(X, suc(N)) :- X > 0, X0 is X - 1, numc(X0, N).

leq(z, Y).
leq(suc(X), suc(Y)) :- leq(X, Y).

mul(X, z, z).
mul(X, suc(Y), Z) :- mul(X, Y, W), add(X, W, Z).

File: 5. Semester/Deklarativ Programmering/TØ/Week 7/EXL07/lib/Nat.ml
type nat = Z | S of nat

let rec numc x = if x = 0 then Z else S (numc (x - 1))

let rec numf = fun x -> if x = 0 then Z else S (numf (x - 1))

let rec nump = fun x -> if (=) x 0 then Z else S (nump ((-) x 1))
